runtime/server/generate_chart.go

```diff
@@ -0,0 +1,174 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/drivers"
+	"github.com/rilldata/rill/runtime/server/auth"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (s *Server) GenerateChartSpec(ctx context.Context, req *runtimev1.GenerateChartSpecRequest) (*runtimev1.GenerateChartSpecResponse, error) {
+	// TODO: telemetry
+
+	// Must have edit permissions on the repo
+	if !auth.GetClaims(ctx).CanInstance(req.InstanceId, auth.EditRepo) {
+		return nil, ErrForbidden
+	}
+
+	var schema *runtimev1.StructType
+	var table string
+	var err error
+
+	if req.Chart != "" {
+		// Get chart
+		chartSpec, err := s.getChart(ctx, req.InstanceId, req.Chart)
+		if err != nil {
+			return nil, err
+		}
+
+		// Resolve schema of the chart
+		schema, err = s.runtime.ResolveSchema(ctx, &runtime.ResolveOptions{
+			InstanceID:         req.InstanceId,
+			Resolver:           chartSpec.Resolver,
+			ResolverProperties: chartSpec.ResolverProperties.AsMap(),
+			Args:               nil,
+			UserAttributes:     auth.GetClaims(ctx).Attributes(),
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		table = req.Chart // not needed but better not to leave it empty
+	} else if req.Table != "" {
+		// Get instance
+		inst, err := s.runtime.Instance(ctx, req.InstanceId)
+		if err != nil {
+			return nil, err
+		}
+
+		// Connect to connector and check it's an OLAP db
+		handle, release, err := s.runtime.AcquireHandle(ctx, req.InstanceId, inst.ResolveOLAPConnector())
+		if err != nil {
+			return nil, err
+		}
+		defer release()
+		olap, ok := handle.AsOLAP(req.InstanceId)
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "connector is not an OLAP connector")
+		}
+
+		// Get table info
+		tbl, err := olap.InformationSchema().Lookup(ctx, req.Table)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, "table not found")
+		}
+
+		schema = tbl.Schema
+		table = req.Table
+	} else {
+		// TODO: support metrics view
+		return nil, errors.New("at lest one of chart or table must be specified")
+	}
+
+	resp, err := s.generateChartWithAI(ctx, req.InstanceId, table, req.Prompt, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the vega lite spec json to string
+	spec, err := json.Marshal(resp.VegaLiteSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtimev1.GenerateChartSpecResponse{
+		VegaLiteSpec: string(spec),
+		Sql:          resp.SQL,
```

Comment:
If the AI is anyway generating the resolver also, why not just output the whole file here, like for metrics view generation?

Reply:
I am thinking there might be different configuration in the chart that would need to be untouched. Would probably be better for the yaml to be built in the UI in that case.